contracts/Content/UniqueContent.sol

```diff
@@ -0,0 +1,191 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import "@openzeppelin/contracts-upgradeable/token/ERC721/ERC721Upgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/token/ERC1155/utils/ERC1155HolderUpgradeable.sol";
+import "@openzeppelin/contracts-upgradeable/interfaces/IERC2981Upgradeable.sol";
+import "../libraries/LibAsset.sol";
+import "./MultipleRoyalties.sol";
+import "./interfaces/IMultipleRoyalties.sol";
+import "./interfaces/IUniqueContent.sol";
+import "./interfaces/IContent.sol";
+
+contract UniqueContent is IUniqueContent, IMultipleRoyalties, MultipleRoyalties, ERC721Upgradeable, ERC1155HolderUpgradeable, IERC2981Upgradeable {
+
+    /***************** Stored Variables *****************/
+    mapping(uint256 => LibAsset.UniqueAsset) uniqueAssetInfo;
+    uint256 private uniqueIdsCounter;
+    
+    /** Asset Minting
+    * @dev If the royalties are valid and if the caller has the original item in their wallet, it takes the original asset, mints the unique asset and updates mappings
+    * @param _data LibAsset.UniqueAssetCreateData structure object
+    */
+    function mint(LibAsset.UniqueAssetCreateData memory _data) external override {
+        (, uint256 _originalRoyaltyRate) = IERC2981Upgradeable(_data.contentAddress).royaltyInfo(_data.tokenId, 1e6);
+        require(_verifyRoyalties(_data.royaltyReceivers, _data.royaltyRates, _originalRoyaltyRate), "The royalties entered are invalid");
```

Comment:
Always put the require() statements at the top (or as much as possible). This means move the .royaltyInfo() call beneth the requires.

The reason is that we don't want to spend gas and make the royaltyInfo() call if it'll fail anyway with the require() calls. It's a waste of gas

Reply:
I need to call royaltyInfo before verifyRoyalties because the verifyRoyalties require statement needs the value of the originalRoyaltyRate to make sure that the royalties entered in addition to the original item's royalty doesn't exceed 1e6. I can put the balanceOf require statement above royaltyInfo though.

Reply:
okay, yeah, this is fine.